### PR TITLE
fix: Allow requesting scalar values in workspace/configuration

### LINKED ISSUE #648: workspace/configuration does not work when scalar value is requested
When using the language server for the [slang shading language](https://github.com/shader-slang/slang) It will send a workspace/configuration request. The sections requested there point to "scalar" values. They are not json objects but strings and bools directly.

In VSCode this will work correctly. But with this plugin the return value is always null (tested with a7d451d). 

Messages in VSCode:
Image

Messages with this plugin:
Image

Configuration of the lsp in this plugin:
```{
    "slang": {
        "predefinedMacros": [],
        "additionalSearchPaths": [],
        "searchInAllWorkspaceDirectories": true,
        "enableCommitCharactersInAutoCompletion": "membersOnly",
        "format": {
            "clangFormatLocation": "",
            "clangFormatStyle": "file",
            "clangFormatFallbackStyle": "{BasedOnStyle: Microsoft, BreakBeforeBraces: Allman, ColumnLimit: 0}",
            "allowLineBreakChangesInOnTypeFormatting": false,
            "allowLineBreakChangesInRangeFormatting": false
        },
        "inlayHints": {
            "deducedTypes": true,
            "parameterNames": true
        }
    },
    "slangLanguageServer": {
        "trace": {
            "server": "verbose"
        }
    }
}

### PR DESCRIPTION
**Pull request by JanThomas118**
This is my attempt to fix #648 

My knowledge about LSP and its different language servers is very limited. Atleast for the faulty request shown in the issue and the configuration listed there this will now return the correct values.

### CODE REVIEW

**Comment by fbricon** [comment]
We'll likely need a new method, as I believe this one is API, changing signature would break 3rd party adopters. @angelozerr ?

**Comment by angelozerr** [comment]
@fbricon I think nobody overrides this method and use the default code.

But I think we should have tests now with this findSettings method.

@JanThomas118 do you think you could write them? If yes I suggest that you create a SettingsHelper class which have a static findSentings method.

You could after that use it in the current code and in the test.

**Comment by JanThomas118** [comment]
Good point. My change is changing public API i actually did not think about that. But the parameter could just continue being JsonObject like before. Then the signature would be identical to before.

Regarding the tests: I can try to do that. This is the first time i am writing java code in years so. I will try my best

File: src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java [hunk 1]
```diff
@@ -0,0 +1,23 @@
+package com.redhat.devtools.lsp4ij.client;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+public class SettingsHelper {
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java:6]: Add javdoc

File: src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java [hunk 2]
```diff
@@ -0,0 +1,23 @@
+package com.redhat.devtools.lsp4ij.client;
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java:1]: Add header

File: src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java [hunk 3]
```diff
@@ -0,0 +1,23 @@
...
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+
+public class SettingsHelper {
+
+    private SettingsHelper() {
+    }
+
+    public static JsonElement findSettings(String[] sections, JsonObject jsonObject) {
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java:11]: Add javdoc

Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java:11]: Add @NotNull for sections and @Nullable I think for jsonObject

Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java:11]: Rename jsonObject with parent

File: src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+        JsonElement current = jsonObject;
+        for (String section : sections) {
+            if (current instanceof JsonObject object) {
```
Comment (angelozerr) [inline src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java:14]: Rename object with child

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java [hunk 5]
```diff
@@ -0,0 +1,73 @@
+package com.redhat.devtools.lsp4ij.client;
+
+import com.google.gson.JsonElement;
+import com.google.gson.JsonObject;
+import com.google.gson.JsonParser;
+import com.intellij.testFramework.fixtures.BasePlatformTestCase;
+
+import java.io.StringReader;
+
+public class SettingsHelperTest extends BasePlatformTestCase {
```
Comment (angelozerr) [inline src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:10]: Add javadoc

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java [hunk 6]
```diff
@@ -0,0 +1,73 @@
+package com.redhat.devtools.lsp4ij.client;
```
Comment (angelozerr) [inline src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:1]: Add header

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java [hunk 7]
```diff
@@ -0,0 +1,73 @@
...
+                    "myobjectsettings": {
+                        "subsettingA": 1,
+                        "subsettingB": 2
+                    }
+                }
+            }
+            """;
+
+    public void testGetSettingsPrimitiveValue() {
+        JsonObject jsonObject = JsonParser.parseReader(new StringReader(testJson)).getAsJsonObject();
```
Comment (angelozerr) [inline src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:25]: To avoid duplicating code I suggest that you move this load and the assert check in one static methad assertFindSettings(String json, String[] sections, String expected)

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java [hunk 8]
```diff
@@ -0,0 +1,73 @@
...
+
+    public void testGetSettingsIdentityOnEmptySections() {
+        JsonObject jsonObject = JsonParser.parseReader(new StringReader(testJson)).getAsJsonObject();
+        String[] requestedPath = new String[0];
+
+        JsonElement result = SettingsHelper.findSettings(requestedPath, jsonObject);
+
+        assertEquals(jsonObject, result);
+    }
+}
```
Comment (angelozerr) [inline src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:73]: Please add findSettings tests with one section and three section (today your test is with zero or two sections)

**Comment by fbricon** [comment]
> @fbricon I think nobody overrides this method and use the default code.

We have no way to be sure. We should just be conservative about it

**Comment by angelozerr** [comment]
Ok lets keep the original signature.

File: src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java (hunk 4 above)
Comment (JanThomas118) [reply src/main/java/com/redhat/devtools/lsp4ij/client/SettingsHelper.java]: I agree that "object" was a terrible name. Child would not be a correct name though because the actual child is retrieved with get() one line below. This variable is just "current" casted to JsonObject. I therefore called it currentObject

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java (hunk 8 above)
Comment (angelozerr) [reply src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:88]: @JanThomas118 is there any chance that you add those tests?

Comment (angelozerr) [reply src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:88]: It would be nice also to add test with zero and for sections to cover all usecases.

@JanThomas118 if you have no time to do that, please tell me. We could merge your PR and I will add tests.

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java (hunk 7 above)
Comment (JanThomas118) [reply src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java]: As far as i understood your suggestion this has now been done. Correct me of i am wrong

File: src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java (hunk 8 above)
Comment (JanThomas118) [reply src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:88]: Zero sections is handled by testGetSettingsIdentityOnEmptySections. Currently there are tests for selections with length 0 1 2 and 3. If you want one with four i can add that but what benefit do you hope this brings when 2 and 3 is already covered?

Comment (angelozerr) [reply src/test/java/com/redhat/devtools/lsp4ij/client/SettingsHelperTest.java:88]: I suggested to add  four to cover the case where sections.length > flattered JSON properties lengh.

But it is OK.
